pool/swap_test.gno

```diff
@@ -0,0 +1,676 @@
+			pool: &Pool{
+				balances: Balances{
+					token0: u256.NewUint(1000),
+					token1: u256.NewUint(2000),
+				},
+			},
+			from:         testutils.TestAddress("from_addr"),
+			to:           testutils.TestAddress("to_addr"),
+			tokenPath:    "token0_path",
+			amount:       i256.NewInt(0),
+			isToken0:     true,
+			expectedBal0: u256.NewUint(1000), // unchanged
+			expectedBal1: u256.NewUint(2000), // unchanged
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// mock transferFromByRegisterCall
+			oldTransferFromByRegisterCall := transferFromByRegisterCall
+			defer func() { transferFromByRegisterCall = oldTransferFromByRegisterCall }()
+
+			transferFromByRegisterCall = func(tokenPath string, from, to std.Address, amount uint64) bool {
+				// mock the transfer (just return true)
+				return true
+			}
+
+			tt.pool.transferFromAndVerify(tt.from, tt.to, tt.tokenPath, u256.MustFromDecimal(tt.amount.ToString()), tt.isToken0)
+
+			if !tt.pool.balances.token0.Eq(tt.expectedBal0) {
+				t.Errorf("token0 balance mismatch: expected %s, got %s",
+					tt.expectedBal0.ToString(),
+					tt.pool.balances.token0.ToString())
+			}
+
+			if !tt.pool.balances.token1.Eq(tt.expectedBal1) {
+				t.Errorf("token1 balance mismatch: expected %s, got %s",
+					tt.expectedBal1.ToString(),
+					tt.pool.balances.token1.ToString())
+			}
+		})
+	}
+
+	t.Run("negative value handling", func(t *testing.T) {
+		pool := &Pool{
+			balances: Balances{
+				token0: u256.NewUint(1000),
+				token1: u256.NewUint(2000),
+			},
+		}
+
+		oldTransferFromByRegisterCall := transferFromByRegisterCall
+		defer func() { transferFromByRegisterCall = oldTransferFromByRegisterCall }()
+
+		transferFromByRegisterCall = func(tokenPath string, from, to std.Address, amount uint64) bool {
+			return true
+		}
+
+		negativeAmount := i256.NewInt(-500)
+		pool.transferFromAndVerify(
+			testutils.TestAddress("from_addr"),
+			testutils.TestAddress("to_addr"),
+			"token0_path",
+			u256.MustFromDecimal(negativeAmount.Abs().ToString()),
+			true,
+		)
+
+		expectedBal := u256.NewUint(1500) // 1000 + 500 (absolute value)
+		if !pool.balances.token0.Eq(expectedBal) {
+			t.Errorf("negative amount handling failed: expected %s, got %s",
+				expectedBal.ToString(),
+				pool.balances.token0.ToString())
+		}
+	})
+
+	t.Run("uint64 overflow value", func(t *testing.T) {
+		pool := &Pool{
+			balances: Balances{
+				token0: u256.NewUint(1000),
+				token1: u256.NewUint(2000),
+			},
+		}
+
+		hugeAmount := i256.FromUint256(u256.MustFromDecimal("18446744073709551616")) // 2^64
+
+		defer func() {
+			if r := recover(); r == nil {
+				t.Error("expected panic for amount exceeding uint64 range")
+			}
+		}()
+
+		pool.transferFromAndVerify(
+			testutils.TestAddress("from_addr"),
+			testutils.TestAddress("to_addr"),
+			"token0_path",
+			u256.MustFromDecimal(hugeAmount.ToString()),
+			true,
+		)
+	})
+}
+
+func TestDrySwap(t *testing.T) {
```

Comment:
Can you add tests with zero amounts (`amountSpecified: "0"`) and insufficient balances? 
These two edge cases must be tested in `DrySwap` and in the actual `Swap`.

Reply:
@notJoon @onlyhyde How can we make sure that `Swap` and `DrySwap` have the same output? can we add a test for this?

Reply:
@dongwon8247 
added [this commit ](https://github.com/gnoswap-labs/gnoswap/pull/421/commits/5a924a9acd6a57e91660d2256ec424da9a1ed8f3)

Both swap and dryswap use the same function. The computation for the actual swap is done using `computeSwap` function